src/detect-ssh-hassh-server.c

```diff
@@ -0,0 +1,205 @@
+#include "rust.h"
+
+
+#define KEYWORD_NAME "hasshServer"
+#define KEYWORD_NAME_LEGACY "hasshServer"
+#define KEYWORD_DOC "hassh-keywords.html#hasshServer"
+#define BUFFER_NAME "hasshServer"
+#define BUFFER_DESC "Ssh Client Fingerprinting For Ssh Servers"
+static int g_ssh_hassh_buffer_id = 0;
+
+
+static InspectionBuffer *GetSshData(DetectEngineThreadCtx *det_ctx,
+        const DetectEngineTransforms *transforms, Flow *_f,
+        const uint8_t flow_flags, void *txv, const int list_id)
+{
+    
+    SCEnter();
+
+    InspectionBuffer *buffer = InspectionBufferGet(det_ctx, list_id);
+
+    if (buffer->inspect == NULL) {
+        const uint8_t *hasshServer = NULL;
+        uint32_t b_len = 0;
+
+        if (rs_ssh_tx_get_hassh_server(txv, &hasshServer, &b_len, flow_flags) != 1)
+            return NULL;
+        if (hasshServer == NULL || b_len == 0) {
+            SCLogDebug("SSH hassh not set");
+            return NULL;
+        }
+
+        InspectionBufferSetup(buffer, hasshServer, b_len);
+        InspectionBufferApplyTransforms(buffer, transforms);
+    }
+
+    return buffer;
+}
+
+/**
+ * \brief this function setup the hasshServer modifier keyword used in the rule
+ *
+ * \param de_ctx Pointer to the Detection Engine Context
+ * \param s      Pointer to the Signature to which the current keyword belongs
+ * \param str    Should hold an empty string always
+ *
+ * \retval 0  On success
+ * \retval -1 On failure
+ * \retval -2 on failure that should be silent after the first
+ */
+static int DetectSshHasshServerSetup(DetectEngineCtx *de_ctx, Signature *s, const char *arg)
+{
+    if (DetectBufferSetActiveList(s, g_ssh_hassh_buffer_id) < 0)
+        return -1;
+
+    if (DetectSignatureSetAppProto(s, ALPROTO_SSH) < 0)
+        return -1;
+            
+    /* try to enable Hassh */
+    SSHEnableHassh();
+
+    /* Check if Hassh is disabled */
+    if (!RunmodeIsUnittests() && !SSHHasshIsEnabled()) {
+        if (!SigMatchSilentErrorEnabled(de_ctx, DETECT_AL_SSH_HASSH_SERVER)) {
+            SCLogError(SC_WARN_HASSH_DISABLED, "hassh support is not enabled");
+        }
+        return -2;
+    }
+
+    return 0;
+
+}
+
+
+static _Bool DetectSshHasshServerHashValidateCallback(const Signature *s,
+                                              const char **sigerror)
+{
+    const SigMatch *sm = s->init_data->smlists[g_ssh_hassh_buffer_id];
+    for ( ; sm != NULL; sm = sm->next)
+    {
+        if (sm->type != DETECT_CONTENT)
+            continue;
+
+        const DetectContentData *cd = (DetectContentData *)sm->ctx;
+
+        if (cd->flags & DETECT_CONTENT_NOCASE) {
+            *sigerror = "hasshServer should not be used together with "
+                        "nocase, since the rule is automatically "
+                        "lowercased anyway which makes nocase redundant.";
+            SCLogWarning(SC_WARN_POOR_RULE, "rule %u: %s", s->id, *sigerror);
+        }
+
+        if (cd->content_len == 32)
+            return TRUE;
+
```

Comment:
Should we check that we have hexadecimal characters as well ?

Reply:
```suggestion
if (cd->content_len != 32)
        {
            *sigerror = "Invalid length of the specified hassh (should "
                        "be 32 characters long). This rule will therefore "
                        "never match.";
            SCLogWarning(SC_WARN_POOR_RULE,  "rule %u: %s", s->id, *sigerror);
            return FALSE;
        }
        for (size_t i = 0; i < cd->content_len; ++i)
        {
            if(!isxdigit(cd->content[i])) 
            {
                *sigerror = "Invalid hassh string (should be string of hexademical characters)."
                            "This rule will therefore never match.";
                SCLogWarning(SC_WARN_POOR_RULE,  "rule %u: %s", s->id, *sigerror);
                return FALSE;
            }
        }
```